common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1908 @@
+	if ms.chainID.String() != chainID.String() {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: %w", ErrInvalidChainID)
+	}
+	ms.addressStatesLock.Lock()
+	defer ms.addressStatesLock.Unlock()
+	as, ok := ms.addressStates[fromAddress]
+	if !ok {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: %w", ErrAddressNotFound)
+	}
+
+	// ensure that the address is not already busy with a transaction in progress
+	if as.inprogress != nil {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: address %s is already busy with a transaction in progress", fromAddress)
+	}
+
+	etx, err := as.PeekNextUnstartedTx()
+	if err != nil || etx == nil {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: %w", err)
+	}
+	tx = ms.deepCopyTx(*etx)
+
+	return nil
+}
+
+// SaveReplacementInProgressAttempt saves a replacement attempt for a transaction that is in_progress.
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) SaveReplacementInProgressAttempt(
+	ctx context.Context,
+	oldAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	replacementAttempt *txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+) error {
+	if oldAttempt.State != txmgrtypes.TxAttemptInProgress || replacementAttempt.State != txmgrtypes.TxAttemptInProgress {
+		return fmt.Errorf("save_replacement_in_progress_attempt: expected attempts to be in_progress")
+	}
+	if oldAttempt.ID == 0 {
+		return fmt.Errorf("save_replacement_in_progress_attempt: expected oldattempt to have an ID")
+	}
+
+	ms.addressStatesLock.Lock()
+	defer ms.addressStatesLock.Unlock()
+	as, ok := ms.addressStates[oldAttempt.Tx.FromAddress]
+	if !ok {
+		return fmt.Errorf("save_replacement_in_progress_attempt: %w", ErrAddressNotFound)
+	}
+
+	// Persist to persistent storage
+	if err := ms.txStore.SaveReplacementInProgressAttempt(ctx, oldAttempt, replacementAttempt); err != nil {
+		return fmt.Errorf("save_replacement_in_progress_attempt: %w", err)
+	}
+
+	tx, err := as.PeekInProgressTx()
+	if tx == nil {
+		return fmt.Errorf("save_replacement_in_progress_attempt: %w", err)
+	}
+	// TODO: DOES THIS ATTEMPT HAVE AN ID? IF NOT, HOW DO WE GET IT?
+	tx.TxAttempts = []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{*replacementAttempt}
+
+	return nil
+}
+
+// UpdateTxFatalError updates a transaction to fatal_error.
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) UpdateTxFatalError(ctx context.Context, tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
+	if tx.State != TxInProgress && tx.State != TxUnstarted {
+		return fmt.Errorf("update_tx_fatal_error: can only transition to fatal_error from in_progress, transaction is currently %s", tx.State)
+	}
+	if !tx.Error.Valid {
+		return fmt.Errorf("update_tx_fatal_error: expected error field to be set")
+	}
+
+	ms.addressStatesLock.Lock()
```

Comment:
`ms.addressStatesLock.Lock()` blocks operations for all addresses. This might lead to performance degradation as it's not possible to perform concurrent operations with different addresses. In all operations, except close, we are reading `addressStates`, so it's sufficient to do `RLock` on `RWLock`